### PR TITLE
Introducing the new pipeline 

### PR DESCRIPTION
**Pull request by johnsimons**
This PR introduces the new Calamari pipeline that is being used in Sashimi/Calamari modules.

The new pipeline streamlines the work a developer needs to perform when implementing a Calamari command.
Here is an example of a before and after makeover for `DeployAzureWebCommand`:

```c#
public override int Execute(string[] commandLineArguments)
{
    Options.Parse(commandLineArguments);

    Guard.NotNullOrWhiteSpace(pathToPackage,
        "No package file was specified. Please pass --package YourPackage.nupkg");

    if (!File.Exists(pathToPackage))
        throw new CommandException("Could not find package file: " + pathToPackage);

    Log.Info("Deploying package:    " + pathToPackage);

    var fileSystem = new WindowsPhysicalFileSystem();
    var replacer = new ConfigurationVariablesReplacer(variables, log);
    var structuredConfigVariableReplacer = new StructuredConfigVariableReplacer(new JsonFormatVariableReplacer(fileSystem, log), new YamlFormatVariableReplacer(), log);
    var transformFileLocator = new TransformFileLocator(fileSystem, log);
    var structuredConfigVariablesService = new StructuredConfigVariablesService(structuredConfigVariableReplacer, fileSystem, log);

    var conventions = new List<IConvention>
    {
        new DelegateInstallConvention(d => extractPackage.ExtractToStagingDirectory(pathToPackage)),
        new ConfiguredScriptConvention(new ConfiguredScriptBehaviour(DeploymentStages.PreDeploy, log, fileSystem, scriptEngine, commandLineRunner)),
        new PackagedScriptConvention(new PackagedScriptBehaviour(log, DeploymentStages.PreDeploy, fileSystem, scriptEngine, commandLineRunner)),
        new DelegateInstallConvention(d => substituteInFiles.SubstituteBasedSettingsInSuppliedVariables(d)),
        new ConfigurationTransformsConvention(new ConfigurationTransformsBehaviour(fileSystem, variables, transformFileLocator, log)),
        new ConfigurationVariablesConvention(new ConfigurationVariablesBehaviour(fileSystem, replacer, log)),
        new JsonConfigurationVariablesConvention(new JsonConfigurationVariablesBehaviour(structuredConfigVariablesService)),
        new PackagedScriptConvention(new PackagedScriptBehaviour(log, DeploymentStages.Deploy, fileSystem, scriptEngine, commandLineRunner)),
        new ConfiguredScriptConvention(new ConfiguredScriptBehaviour(DeploymentStages.Deploy, log, fileSystem, scriptEngine, commandLineRunner)),
        new AzureWebAppConvention(log),
        new LogAzureWebAppDetails(log),
        new PackagedScriptConvention(new PackagedScriptBehaviour(log, DeploymentStages.PostDeploy, fileSystem, scriptEngine, commandLineRunner)),
        new ConfiguredScriptConvention(new ConfiguredScriptBehaviour(DeploymentStages.PostDeploy, log, fileSystem, scriptEngine, commandLineRunner)),
    };

    var deployment = new RunningDeployment(pathToPackage, variables);
    var conventionRunner = new ConventionProcessor(deployment, conventions, log);

    conventionRunner.RunConventions();

    return 0;
}
```
vs
```c#
[Command("deploy-azure-web", Description = "Extracts and installs a deployment package to an Azure Web Application")]
public class DeployAzureWebCommand : PipelineCommand
{
    protected override IEnumerable<IDeployBehaviour> Deploy(DeployResolver resolver)
    {
        yield return resolver.Create<AzureWebAppBehaviour>();
        yield return resolver.Create<LogAzureWebAppDetailsBehaviour>();
    }
}
```

As you can see the end result is a lot shorter.

### The main changes
- There are no `IConventions`, we now use `IBehaviours`;
- The command implements `PipelineCommand` instead of `Command`;
- The entry point has also changed, your `Program.cs` now derives from `CalamariFlavourProgramAsync`, yes the pipeline is all async 🎉 
- The `PipelineCommand` allows the developer to override a series of pipeline "stages", the only "stage" that is mandatory is the `Deploy` as seen on the previous example above. Here is a more complex pipeline for `DeployAzureCloudServiceCommand`:
```c#
[Command("deploy-azure-cloud-service", Description = "Extracts and installs an Azure Cloud-Service")]
public class DeployAzureCloudServiceCommand : PipelineCommand
{
    protected override IEnumerable<IBeforePackageExtractionBehaviour> BeforePackageExtraction(BeforePackageExtractionResolver resolver)
    {
        yield return resolver.Create<SwapAzureDeploymentBehaviour>();
    }

    protected override IEnumerable<IAfterPackageExtractionBehaviour> AfterPackageExtraction(AfterPackageExtractionResolver resolver)
    {
        yield return resolver.Create<FindCloudServicePackageBehaviour>();
        yield return resolver.Create<EnsureCloudServicePackageIsCtpFormatBehaviour>();
        yield return resolver.Create<ExtractAzureCloudServicePackageBehaviour>();
        yield return resolver.Create<ChooseCloudServiceConfigurationFileBehaviour>();
    }

    protected override IEnumerable<IPreDeployBehaviour> PreDeploy(PreDeployResolver resolver)
    {
        yield return resolver.Create<ConfigureAzureCloudServiceBehaviour>();
    }

    protected override IEnumerable<IDeployBehaviour> Deploy(DeployResolver resolver)
    {
        yield return resolver.Create<RePackageCloudServiceBehaviour>();
        yield return resolver.Create<UploadAzureCloudServicePackageBehaviour>();
        yield return resolver.Create<DeployAzureCloudServicePackageBehaviour>();
    }
}
```

### Migration to new pipeline steps:
1. Make sure you are referencing `Calamari.Common` package
1. Update `Program.cs` to implement `CalamariFlavourProgramAsync`
1. the the command you are migrating, instead of implementing `ICommand`, implement `CommandPipeline`
1. Identify all framework conventions in your command and delete them, example
    image
1. Convert each convention to an `IBehaviour`, but beaware, you don't implement `IBehaviour` directly, you need to pick the right stage behaviour
    image
    The migration is very simple, there is only 2 methods to implement, `IsEnabled` is where you check if you need to run based on usually variables set, `Execute` is pretty much what you used to write in the `Install`.

### CODE REVIEW

File: source/Calamari.Common/Plumbing/Pipeline/PipelineCommand.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            catch (Exception installException)
+            {
+                Console.Error.WriteLine("Running rollback behaviours...");
+
+                deployment.Error(installException);
+
+                try
+                {
+                    // Rollback behaviours include tasks like DeployFailed.ps1
+                    await ExecuteBehaviour(deployment, lifetimeScope.Resolve<RollbackScriptBehaviour>());
```
Comment (flin-8) [inline source/Calamari.Common/Plumbing/Pipeline/PipelineCommand.cs:66]: I'm pretty sure only `DeployPackageCommand` and `DeployJavaArchiveCommand` call `RollbackScriptConvention`.

Also, `JavaLibraryCommand`, `DeployPackageCommand` and `DeployJavaArchiveCommand` call `FeatureRollbackConvention`.

We probably need to add a hookpoint for rollbacks?

Comment (johnsimons) [reply source/Calamari.Common/Plumbing/Pipeline/PipelineCommand.cs:66]: We can adjust the pipeline once we tackle these commands

Comment (flin-8) [reply source/Calamari.Common/Plumbing/Pipeline/PipelineCommand.cs:66]: If that's the case, I think we should remove the rollback altogether and put in a `//TODO` so it's explicitly missing

File: source/Calamari.Common/Features/ConfigurationTransforms/TransformFileLocator.cs [hunk 2]
```diff
@@ -169,25 +167,23 @@ static bool DoesFileMatchWildcardPattern(string fileName, string? pattern)
             return Regex.IsMatch(fileName, regexBuilder.ToString(), RegexOptions.IgnoreCase);
         }
 
-        [return: NotNullIfNotNull("path")]
-        static string? GetDirectoryName(string? path)
+        static string GetDirectoryName(string path)
```
Comment (flin-8) [inline source/Calamari.Common/Features/ConfigurationTransforms/TransformFileLocator.cs:170]: double check if all these changes are needed/intentional

File: source/Calamari.Common/Features/Scripting/PackagedScriptRunner.cs [hunk 3]
```diff
@@ -62,9 +63,9 @@ string FindPreferredScript(RunningDeployment deployment)
             var supportedScriptExtensions = scriptEngine.GetSupportedTypes();
 
             var files = (from file in FindScripts(deployment)
-                         let preferenceOrdinal = Array.IndexOf(supportedScriptExtensions, file.ToScriptType())
-                         orderby preferenceOrdinal
-                         select file).ToArray();
+                let preferenceOrdinal = Array.IndexOf(supportedScriptExtensions, ScriptTypeExtensions.ToScriptType((string) file))
```
Comment (flin-8) [inline source/Calamari.Common/Features/Scripting/PackagedScriptRunner.cs:66]: this line can be reverted

File: source/Calamari.Common/Plumbing/Pipeline/PipelineCommand.cs (hunk 1 above)
Comment (flin-8) [reply source/Calamari.Common/Plumbing/Pipeline/PipelineCommand.cs:66]: OK, so this is tied in with the ConfiguredScriptBehaviour

**Pull request merged**